### DISCUSSION TITLE
Are large grids a problem in mesa?

### DISCUSSION THREAD

**Discussion by CRG180** [mesa/mesa#1931] [Q&A]
Are large grids computationally burdensome?  When building a 10000 x 10000 grid it takes about 1 min to build the grid.  Is this normal? I suppose 10000^2 is a large number.  My initial thought is that I would like to build the grid with 1 x 1-meter resolution.  I'm I missing something? What are some best practices for sizing the grid? 
 
`class CombatModel(mesa.Model):
    
    def __init__(self, width = 10000, height = 10000):
        self.schedule = mesa.time.RandomActivation(self)
        self.grid =mesa.space.MultiGrid(height, width, False)
        self.step_num = 1 `

**Comment by rht** [comment]
You may use the recently merged new feature `PropertyLayer`, #1898, to scale operations on large grid. It's just that while the operations on the `PropertyLayer` may be fast, the operations on the 2D grid itself is still slow. If we could migrate the 2D grid implementation to be based on NumPy array, then we could scale to large size of the grid.

**Reply by quaquel** [reply]
I ran an experiment over the weekend where I reimplemented _Grid and SingleGrid using NumPy instead of the current list of lists. It was not notably faster than the current implementation. The main reason for this is that the dtype of the NumPy array is object (to be able to store agent instances). Over the coming days, I hope to have time for a larger series of experiments comparing the current list of lists to NumPy arrays with unique_id integers/cell objects with unique integers.

**Reply by rht** [reply]
Implementing the `_Grid` an array of unique_id integers is going to have the NumPy. We used to have similar experiment, but in Cython instead: https://github.com/projectmesa/mesa/discussions/1583. Cython's typed memoryview is essentially NumPy array on steroids. There was a poll on reimplementing space.py in Cython: https://github.com/projectmesa/mesa/discussions/1610.

**Reply by rht** [reply]
https://github.com/rht/mesa-perf has a more detailed benchmark on the Cython speedup for each `_Grid` operations.

**Reply by quaquel** [reply]
Thanks for that!

So, if I want specifically to contribute to moving the cython version into the main repo (which has majority support atm in the vote), where could I start?

**Reply by Corvince** [reply]
> I ran an experiment over the weekend where I reimplemented _Grid and SingleGrid using NumPy instead of the current list of lists. It was not notably faster than the current implementation. The main reason for this is that the dtype of the NumPy array is object (to be able to store agent instances). Over the coming days, I hope to have time for a larger series of experiments comparing the current list of lists to NumPy arrays with unique_id integers/cell objects with unique integers.

I wouldn't invest too much time into this. If you think about it, what are the possible performance gains? Propertylayer is orders of magnitude faster because you can vectorize its operations, for example increment all values almost "at once". But for an ordinary agent grid, you don't "do" that much. One thing you can do is retrieve agents from a set of coordinates. There is not much of a difference between a list or a numpy array here, especially if you then need to translate from ids to actual agents. Another thing is neighborhood lookup. The main slowdowns here are bounds checks and/or wrapping. But in any way we cache neighborhoods, so this hasn't been a bottleneck in some tim (unless your neighborhoods are dynamic). Moving agents from one position to another also shouldn't be faster with numpy (why would it be?). 

But I really don't want to discourage you, I am indeed very curious of your results. 

Regarding Cython. Is there a way to only cythonize single functions? From the performance benchmarks I think only `get_cell_list_contents   ` seems worth pursuing and it is also the most used function, so that would make sense in my opinion.

**Reply by CRG180** [reply]
Really interesting discussion. For my little test project, I used the continuous grid which is fast. When I get some time to experiment  I want to add an elevation terrain PropertyLayer or an adaption that interpolates between sparse data points, I would then like to create additional methods for line-of-sight calculations between Agents.

I feel like for large regions continuous grids work well. And  PropertyLayers can provide additional data about the enviroment.  What do you gain with the square/ hex grid system?

**Reply by rht** [reply]
> And PropertyLayers can provide additional data about the enviroment. What do you gain with the square/ hex grid system?

AFAIK the PropertyLayer is specific for square grid only, as of currently.

**Reply by rht** [reply]
> So, if I want specifically to contribute to moving the cython version into the main repo (which has majority support atm in the vote), where could I start?

There shouldn't be any blocker. The _Grid hasn't been changed for a while, so a significant portion of https://github.com/rht/mesa-perf should still be relevant

> Regarding Cython. Is there a way to only cythonize single functions? From the performance benchmarks I think only get_cell_list_contents    seems worth pursuing and it is also the most used function, so that would make sense in my opinion.

If you look at the speedups in the mesa-perf readme, there are a sizeable number of >2x speedups, and some are even >3x.
The most significant speedup:
| method name                | speed-up singlegrid | speed-up multigrid |
| get_cell_list_contents     | 10.23x              | 11.36x             |
| __getitem__ single tuple   | 17.04x              | 17.73x             |

**Reply by Corvince** [reply]
>  | **getitem** single tuple | 17.04x | 17.73x |

Could you explain that one in particular? I was curious and looked at mesa-perf, but the getitem function appears to be unchanged in mesa-perf!? Where does the speedup come from?

**Reply by rht** [reply]
Can't remember exactly. Maybe @Tortar still remembers the explanation.

**Reply by quaquel** [reply]
Just a hunch, but it might be that you lose some of the type-checking overhead by doing getitem in C instead of in python.

**Reply by ameligrana** [reply]
Hi guys,

I don't remember exactly why, but you could try to inspect the annotate cython code to understand it I think

**Reply by Corvince** [reply]
Thanks for jumping in, but I don't find anything that convinces me a 18x speed up would be possible. There are no type annotations so I don't know how cython would do this. Makes me skeptical of the rest of the benchmark!

**Reply by ameligrana** [reply]
I'm not sure you mean the same thing as me, by annotation I mean the ones generated when you compile the files with cython and you pass the necessary compiler flag. Also, I think the benchmark code is also on the repo so you could look at it and see if there are any flaws

**Comment by quaquel** [comment]
> Regarding Cython. Is there a way to only cythonize single functions? From the performance benchmarks I think only `get_cell_list_contents ` seems worth pursuing and it is also the most used function, so that would make sense in my opinion.

Yes it is quite easy to cythonize individual functions. In fact, that is how pandas slowly moved most of their code from python to cython in the days before pandas 1.0.

**Comment by rht** [comment]
I'm currently trying to integrate mesa-perf into projectmesa/mesa, via https://pypi.org/project/hatch-cython/0.0.1/.

**Reply by quaquel** [reply]
What exactly do you have in mind? Something like a github action that allows us to easily see the performance consequences of PRs? To be clear: performance in my view is a relevant but not dominating consideration when considering PRs.

**Reply by rht** [reply]
As in, Cythonizing `_Grid`, starting with `get_cell_list_contents`, until all of `_Grid` gets Cythonized.

**Reply by quaquel** [reply]
I misunderstood, and thanks for clarifying. 

Does anyone think that having a perf GitHub Action would be worthwhile? cc @Corvince, @rht @EwoutH @jackiekazil

**Reply by EwoutH** [reply]
I did some work on automated perf testing when working on [dav1d](https://code.videolan.org/videolan/dav1d). It’s certainly possible, but you need a very stable and consistent system. The GitHub-provided runners for Actions vary wildly, between CPU architecture, clock speed, hyper threading, etc.

So you want to provide your own dedicated system, with the minimum amount of noise possible (disabling hyper-treading, locking clock speeds (so disabling turbo boosting).

This might be a very nice and useful project (bsc/msc thesis? Capita selecta?) for a computer science student. It includes lot of components, from knowing your hardware, ensuring statistical consistency and reporting significance, to UI design, generating automated reports and CI automation.

**Reply by quaquel** [reply]
I was afraid that something like this was the case. So automating it is non-trivial. What about a more modest proposal to include part of https://github.com/JuliaDynamics/ABMFrameworksComparison into the MESA repository (but exclude it from builds and distributions)?

I would like to see a simple way to quickly compare different implementations in terms of their performance at the aggregate level. Small parts of the code are easily tested using perf_counter and the timeit magic. However, it is difficult to anticipate how this scales to complete models.

**Reply by EwoutH** [reply]
An initial step would be creating a `benchmarks.py` script that you can run on your own computer. It will at least be able to see the large performance differences. Then everyone can run it themselves on the main branch and on their PR branch and compare the results.

**Reply by rht** [reply]
Regarding with the Cythonizing, I tried https://pypi.org/project/hatch-cython/0.0.1/#description and it seems to be buggy. The only way forward I can think of is to migrate from Hatch to either `meson-python` (used by pandas and NumPy), or Poetry. Which one do you guys prefer and why? I don't know of a major Cythonized project that uses Poetry. But I know several that still uses setuptools.

**Reply by quaquel** [reply]
I would follow pandas and numpy. It will give us good templates for actions etc. as well.

**Reply by Corvince** [reply]
@EwoutH Good heads-up on runner variability. But as long as the both runs happen on the same action it should be fine, right? I mean for example to compare without a PR and with a PR. The times won't be comparable between Action runs, but at least we would know how it affects performance. But I think it should be an optional Action, only activated when needed.

**Reply by EwoutH** [reply]
Main problem is modern hyperthreading and turboing. While great for optimizing overal performance, they are not good for consistency. On these kind of service, you won't be the only job running on that core.

If the core is busy on the first half and less busy on the second half, that can seriously affect you numbers.
